src/ssim.jl

```diff
@@ -0,0 +1,121 @@
+"""
+    SSIM([kernel], [(α, β, γ)]) <: FullReferenceIQI
+    assess(iqi::SSIM, img, ref)
+    ssim(img, ref)
+
+Structural similarity (SSIM) index is an image quality assessment method based
+on degradation of structural information.
+
+The SSIM index is composed of three components: luminance similarity, contrast
+similarity, and structural similarity. `ssim = 𝐿ᵅ * 𝐶ᵝ * 𝑆ᵞ`,
+where `W := (α, β, γ)` controls relative importance of each components.
+By default `W = (1.0, 1.0, 1.0)`.
+
+In practice, a local version SSIM is used. At each pixel, SSIM is calculated
+locally with neighborhoods weighted by guassian `kernel`, returning a ssim map;
+`ssim` is actaully `mean(ssim_map)`.
+By default `kernel = KernelFactors.gaussian(1.5, 11)`.
+
+!!! info
+
+    RGB images are treated as 3d Gray images, general Color3 images are
+    converted to RGB images first.
+
+# Example
+
+In most cases, `ssim(img, ref)` should be sufficient, which calls the default
+SSIM. One could also instantiate a SSIM and pass it to `assess`, for example:
+
+```julia
+iqi = SSIM(KernelFactors.gaussian(2.5, 17), (1.0, 1.0, 2.0))
+assess(iqi, img, ref)
+```
+
+# Reference
+
+[1] Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P. (2004). Image quality assessment: from error visibility to structural similarity. _IEEE transactions on image processing_, 13(4), 600-612.
+
+[2] Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P. (2003). The SSIM Index for Image Quality Assessment. Retrived May 30, 2019, from http://www.cns.nyu.edu/~lcv/ssim/
+"""
+struct SSIM <: FullReferenceIQI
+    kernel::OffsetArray
+    W::NTuple{3}
+    function SSIM(kernel, W)
+        ndims(kernel) == 1 || throw(ArgumentError("only 1-d kernel is valid"))
+        all(W .>= 0) || throw(ArgumentError("W should be non-negative"))
+        new(centered(kernel), W)
```

Comment:
I use `centered(kernel)` to support both `Array` and `OffsetArray` inputs.
Intuitively, `centered(kernel::OffsetArray) == kernel` should be true.

This behavior is blocked by https://github.com/JuliaImages/ImageFiltering.jl/pull/104

Reply:
I'm not very familiar with the needs here, but just on general terms automatically calling `centered` can be problematic. For instance, I can execute a translation of two pixels by filtering with `[0, 0, 1]`. In that case my lack of `OffsetArray` wrapper is deliberate, and I would be annoyed if someone centered it and turned this into a translation by 1 pixel.

That may not be relevant to the case here, but just pointing out the hazards in "I'm going to assume you've made a mistake and correct it for you" code.

Reply:
That's definitely worth pointing out, thanks!

In ssim context, actually, `kernel` is used to calculate weighted-version mean/std/covariance of patch `Xᵢ`, which consists of pixel `i` and its 5*5 neighbor (assuming the kernel is of length 11). It would be meaningless having a nonsymmetric kernel in this context.

Take the example of `mean`, we use convolution `imfilter` to accelerate the calculation of `μxᵢ = mean(kernel .* Xᵢ)`. And since it's symmetric, `imfilter(x, kernel) = imfilter(x, reflect(kernel))`

I'll change it to throw a warning if `kernel` is not symmetric, that looks better.